packages/ddp-server/livedata_server.js

```diff
@@ -762,16 +762,43 @@ Object.assign(Session.prototype, {
           }
         }
 
-        resolve(DDPServer._CurrentWriteFence.withValue(
-          fence,
-          () => DDP._CurrentMethodInvocation.withValue(
-            invocation,
-            () => maybeAuditArgumentChecks(
-              handler, invocation, msg.params,
+        const getCurrentMethodInvocationResult = () => {
+          const currentContext = DDP._CurrentMethodInvocation.setNewContextAndGetCurrent(
+            invocation
+          );
+
+          try {
+            let result;
+            const resultOrThenable = maybeAuditArgumentChecks(
+              handler,
+              invocation,
+              msg.params,
               "call to '" + msg.method + "'"
-            )
-          )
-        ));
+            );
+            const isThenable =
+              resultOrThenable && typeof resultOrThenable.then === 'function';
+            if (isThenable) {
+              result = Promise.await(resultOrThenable);
+            } else {
+              result = resultOrThenable;
+            }
+            return result;
+          } finally {
+            DDP._CurrentMethodInvocation.set(currentContext);
+          }
+        };
+        const getCurrentWriteFenceResult = () => {
+          const currentContext = DDPServer._CurrentWriteFence.setNewContextAndGetCurrent(
+            fence
+          );
+          try {
+            return getCurrentMethodInvocationResult();
+          } finally {
+            DDPServer._CurrentWriteFence.set(currentContext);
+          }
```

Comment:
That's exactly the same as
```ts
return DDPServer._CurrentWriteFence.withValue(fence, getCurrentMethodInvocationResult);
```
isn't it?

Similarly, the `getCurrentMethodInvocationResult` could use `withValue` too.

Reply:
About the DDPServer._CurrentWriteFence.withValue you're completely right. Fixed.

About the getCurrentMethodInvocationResult, we need to do it this way.

Now, [this function](https://github.com/meteor/meteor/pull/12196/files#diff-173e69ea0353a765b98017d67abc45ec7ce1449178466dc74738324db83f9183R772) can return a promise, and we need to wait for the result.

So the first solution was to create a [`withValueAsync`](https://github.com/meteor/meteor/pull/12196/commits/15b1e6fddfc5a03c4dd17d97d3b577996b9c0c61#diff-47412bfdfb4c9c0628248595b3f6c447bf7ad49535e508628dc704b729028ea8L93-L106) that waits the promise to solve. But [this](https://github.com/meteor/meteor/pull/12196#discussion_r981792165) problem pointed out by Zodern was created. But also by doing his [suggestion](https://github.com/meteor/meteor/pull/12196/commits/15b1e6fddfc5a03c4dd17d97d3b577996b9c0c61#diff-47412bfdfb4c9c0628248595b3f6c447bf7ad49535e508628dc704b729028ea8R93-R106) (wrapping that function in a new Fiber) was changing the dynamics and when you called a `Meteor.call()` without the callback, it didn't return a result anymore. Resulting in a lot of tests breaking. 

So I'm applying there the same solution I did for the [applyAsync](https://github.com/meteor/meteor/pull/12196/files#diff-5a0f1c2953dc6856044b4803bb19c0b301386cd8836c878e36f2776d6316dc79R657-R667). Resolving the promise one level above.